File: Sources/DiskMapCore/ScanEngine.swift
import Foundation

/// Walks a directory tree off the main thread and builds a `FileTree`.
///
/// The walk is `getattrlistbulk`, not `FileManager.enumerator`. The
/// enumerator built a `URL` and a resource-value object per file; on a
/// home folder that was about 373 s. Bulk attributes plus a handful of
/// worker threads is the macOS API for this.
public actor ScanEngine {

    public struct Result: Sendable {
        public var tree: FileTree
        public var itemCount: Int
        public var elapsedSeconds: Double
        /// Max `resident_size` sampled while the walk was still running
        /// (enumerator and per-item `resourceValues` still alive). Not the
        /// process-lifetime `resident_size_max`.
        public var peakResidentBytesDuringWalk: UInt64
        /// `resident_size` after the walk function has returned, so the
        /// enumerator is released and the tree is still retained.
        public var residentBytesAfterEnumeratorRelease: UInt64?
        public var notDownloadedCount: Int
    }

    /// What to store for one enumerated item. Split out so the iCloud
    /// decision can be unit-tested without an evicted file on disk.
    public struct ItemDecision: Sendable, Equatable {
        public var include: Bool
        public var logicalSize: Int64
        public var allocatedSize: Int64
        public var notDownloaded: Bool
        public var skipDescendants: Bool
    }

    public init() {}

    public func scan(root: URL, progress: (@Sendable (Int) -> Void)? = nil) async -> Result {
        let started = ContinuousClock.now
        // `walk` owns the enumerator. Measuring after it returns is the
        // steady state: tree retained, enumerator and per-item
        // resourceValues released. The during-walk peak is sampled only
        // inside `walk` and is not updated here.
        let walked = await Task.detached(priority: .userInitiated) {
            BulkScan.walk(root: root, progress: progress)
        }.value
        var tree = walked.tree
        tree.compact()
        let afterRelease = ProcessMemory.current()
        let elapsed = started.duration(to: .now)
        let result = Result(
            tree: tree,
            itemCount: walked.itemCount,
            elapsedSeconds: seconds(elapsed),
            peakResidentBytesDuringWalk: walked.peakResidentBytesDuringWalk,
            residentBytesAfterEnumeratorRelease: afterRelease?.residentBytes,
            notDownloadedCount: walked.notDownloadedCount
        )
        logSummary(result)
        return result
    }

    /// Symlinks are skipped (the enumerator does not follow them, and
    /// recording the link would double-count whatever it points at).
    /// An evicted iCloud item is recorded, not opened: size keys do not
    /// start a download (see `scan`), so allocated size is the local
    /// footprint (0 when fully evicted) and logical size is the cloud
    /// size. Descendants of a not-downloaded directory are skipped so
    /// listing them can't materialize the folder.
    public static func decide(_ values: URLResourceValues) -> ItemDecision {
        decide(
            isDirectory: values.isDirectory ?? false,
            isSymbolicLink: values.isSymbolicLink ?? false,
            logicalSize: values.fileSize.map(Int64.init),
            allocatedSize: values.totalFileAllocatedSize.map(Int64.init),
            isUbiquitous: values.isUbiquitousItem ?? false,
            downloadingStatusNotDownloaded: values.ubiquitousItemDownloadingStatus == .notDownloaded
        )
    }

    public static func decide(
        isDirectory: Bool,
        isSymbolicLink: Bool,
        logicalSize: Int64?,
        allocatedSize: Int64?,
        isUbiquitous: Bool,
        downloadingStatusNotDownloaded: Bool
    ) -> ItemDecision {
        if isSymbolicLink {
            return ItemDecision(
                include: false,
                logicalSize: 0,
                allocatedSize: 0,
                notDownloaded: false,
                skipDescendants: true
            )
        }

        let notDownloaded = isUbiquitous && downloadingStatusNotDownloaded
        return ItemDecision(
            include: true,
            logicalSize: logicalSize ?? 0,
            allocatedSize: allocatedSize ?? logicalSize ?? 0,
            notDownloaded: notDownloaded,
            skipDescendants: notDownloaded && isDirectory
        )
    }

    private func logSummary(_ result: Result) {
        let after = result.residentBytesAfterEnumeratorRelease.map(String.init) ?? "unavailable"
        let line = "DiskMap scan: items=\(result.itemCount) elapsed=\(String(format: "%.3f", result.elapsedSeconds))s rss_during_walk_peak=\(result.peakResidentBytesDuringWalk) rss_after_enumerator_release=\(after) not_downloaded=\(result.notDownloadedCount)"
        print(line)
        fflush(stdout)
    }

    private func seconds(_ duration: Duration) -> Double {
        let parts = duration.components
        return Double(parts.seconds) + Double(parts.attoseconds) / 1e18
    }
}

File: windows/src/DiskMap.Core/Scan/ScanEngine.cs
using DiskMap.Core.Native;

namespace DiskMap.Core;

/// <summary>
/// Walks a directory tree off the main thread and builds a <see cref="FileTree"/>.
///
/// Primary path is the NTFS MFT (WizTree-style: read $MFT, parse FILE
/// records, rebuild the subtree). Fallback is a FindFirstFileExW work-queue —
/// needed regardless for non-NTFS volumes, non-elevated runs, and network
/// drives. Both produce the same WalkResult.
/// </summary>
public sealed class ScanEngine
{
    public sealed class Result
    {
        public required FileTree Tree { get; init; }
        public required int ItemCount { get; init; }
        public required double ElapsedSeconds { get; init; }
        /// <summary>Max working-set sampled while the walk was still running.</summary>
        public ulong PeakResidentBytesDuringWalk { get; init; }
        /// <summary>Working set after the walk returned, tree retained.</summary>
        public ulong? ResidentBytesAfterWalk { get; init; }
        public int NotDownloadedCount { get; init; }
        /// <summary>"mft" or "win32" — which backend produced this scan.</summary>
        public required string Backend { get; init; }
    }

    /// <summary>
    /// What to store for one enumerated item. Split out so the cloud-
    /// placeholder decision can be unit-tested without a fixture on disk.
    /// </summary>
    public readonly record struct ItemDecision(
        bool Include,
        long LogicalSize,
        long AllocatedSize,
        bool NotDownloaded,
        bool SkipDescendants);

    public async Task<Result> ScanAsync(string root, IProgress<int>? progress = null)
    {
        var started = System.Diagnostics.Stopwatch.StartNew();
        var walked = await Task.Run(() => Walk(root, progress));
        walked.Tree.Compact();
        var afterRelease = ProcessMemory.Current();
        started.Stop();
        var result = new Result
        {
            Tree = walked.Tree,
            ItemCount = walked.ItemCount,
            ElapsedSeconds = started.Elapsed.TotalSeconds,
            PeakResidentBytesDuringWalk = walked.PeakResidentBytesDuringWalk,
            ResidentBytesAfterWalk = afterRelease?.ResidentBytes,
            NotDownloadedCount = walked.NotDownloadedCount,
            Backend = walked.Backend,
        };
        LogSummary(result);
        return result;
    }

    private static WalkResult Walk(string root, IProgress<int>? progress)
    {
        // MFT pays off when the subtree is a large fraction of the volume:
        // it always reads the whole $MFT, so it wins for drive roots and
        // near-root scans (C:\, C:\Users\me — the "scan my drive/profile"
        // cases). Deep subtrees get the proportional FindFirstFileExW walk
        // instead of parsing millions of unrelated records.
        if (DepthFromVolumeRoot(root) <= 2)
        {
            var mft = MftScanner.Walk(root, progress);
            if (mft is not null) return mft;
        }
        return Win32Scanner.Walk(root, progress);
    }

    /// <summary>
    /// Path depth below the volume root: C:\ → 0, C:\Users → 1,
    /// C:\Users\me → 2. Non-local paths (UNC) return int.MaxValue.
    /// </summary>
    internal static int DepthFromVolumeRoot(string path)
    {
        string full = Path.GetFullPath(path).TrimEnd('\\', '/');
        string? root = Path.GetPathRoot(full);
        if (root is null || full.StartsWith(@"\\")) return int.MaxValue;
        var rel = Path.GetRelativePath(root, full);
        if (rel is "." or "") return 0;
        return rel.Split(Path.DirectorySeparatorChar, StringSplitOptions.RemoveEmptyEntries).Length;
    }

    /// <summary>
    /// Reparse points are skipped — but only after the cloud check: OneDrive
    /// placeholders ARE reparse points (tags IO_REPARSE_TAG_CLOUD*), and the
    /// interesting signal is "in the cloud", not "is a reparse point". A
    /// junction to a local dir is skipped like a symlink on macOS — the
    /// target is counted at its real location, never double-counted here.
    ///
    /// A cloud placeholder is recorded, not opened: enumeration does not
    /// recall the content, opening the file would. Allocated size is the
    /// local footprint (0 when fully evicted); logical size is the cloud
    /// size. Descendants of a not-downloaded directory are skipped so
    /// listing them can't materialize the folder.
    /// </summary>
    public static ItemDecision Decide(
        bool isDirectory,
        bool isReparsePoint,
        bool isCloudPlaceholder,
        long logicalSize,
        long allocatedSize)
    {
        if (isReparsePoint && !isCloudPlaceholder)
        {
            return new ItemDecision(
                Include: false, LogicalSize: 0, AllocatedSize: 0,
                NotDownloaded: false, SkipDescendants: true);
        }
        if (isCloudPlaceholder)
        {
            return new ItemDecision(
                Include: true, LogicalSize: logicalSize, AllocatedSize: allocatedSize,
                NotDownloaded: true, SkipDescendants: isDirectory);
        }
        return new ItemDecision(
            Include: true, LogicalSize: logicalSize, AllocatedSize: allocatedSize,
            NotDownloaded: false, SkipDescendants: false);
    }

    /// <summary>
    /// FILE_ATTRIBUTE_* → the decide() inputs. Shared by both scanners.
    /// </summary>
    internal static ItemDecision DecideFromAttributes(
        uint attributes, long logicalSize, long allocatedSize)
    {
        bool isDir = (attributes & Win32.FILE_ATTRIBUTE_DIRECTORY) != 0;
        bool isReparse = (attributes & Win32.FILE_ATTRIBUTE_REPARSE_POINT) != 0;
        bool isCloud = (attributes & (Win32.FILE_ATTRIBUTE_OFFLINE
            | Win32.FILE_ATTRIBUTE_RECALL_ON_OPEN
            | Win32.FILE_ATTRIBUTE_RECALL_ON_DATA_ACCESS)) != 0;
        return Decide(isDir, isReparse, isCloud, logicalSize, allocatedSize);
    }

    private static void LogSummary(Result result)
    {
        string after = result.ResidentBytesAfterWalk?.ToString() ?? "unavailable";
        Console.WriteLine(
            $"DiskMap scan: backend={result.Backend} items={result.ItemCount} " +
            $"elapsed={result.ElapsedSeconds:0.000}s rss_during_walk_peak={result.PeakResidentBytesDuringWalk} " +
            $"rss_after_walk={after} not_downloaded={result.NotDownloadedCount}");
        Console.Out.Flush();
    }
}

/// <summary>Shared output of both scan backends.</summary>
internal sealed class WalkResult
{
    public required FileTree Tree { get; init; }
    public required int ItemCount { get; init; }
    public int NotDownloadedCount { get; init; }
    public ulong PeakResidentBytesDuringWalk { get; init; }
    public required string Backend { get; init; }
}
